presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -48,76 +45,49 @@
-            leakedQueries = ImmutableSet.copyOf(leakedQueryReservations.keySet());
-        }
-    }
+    @GuardedBy("this")
+    private long leakedBytes;
 
     /**
-     *
-     * @param runningQueriesSupplier All running queries on a cluster.
+     * @param queryIdToInfo All queries that the coordinator knows about, along with their optional query info.
      * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
      */
-    void checkForClusterMemoryLeaks(Supplier<Optional<List<QueryId>>> runningQueriesSupplier, Map<QueryId, Long> queryMemoryReservations)
+    void checkForMemoryLeaks(Map<QueryId, Optional<BasicQueryInfo>> queryIdToInfo, Map<QueryId, Long> queryMemoryReservations)
     {
-        requireNonNull(runningQueriesSupplier);
-        requireNonNull(queryMemoryReservations);
-
-        Optional<List<QueryId>> runningQueries = runningQueriesSupplier.get();
-
         Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
                 .stream()
                 .filter(entry -> entry.getValue() > 0)
-                .filter(entry -> !runningQueries.isPresent() || !runningQueries.get().contains(entry.getKey()))
+                .filter(entry -> isLeaked(queryIdToInfo, entry.getKey()))
                 .collect(toImmutableMap(Entry::getKey, Entry::getValue));
 
+        long leakedBytesThisTime = leakedQueryReservations.values().stream().reduce(0L, Long::sum);
         if (!leakedQueryReservations.isEmpty()) {
-            log.debug("Memory leak detected. The following queries are already finished, " +
-                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+            log.warn("Memory leak of %s detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s",
+                    DataSize.succinctBytes(leakedBytes), leakedQueryReservations);
         }
 
         synchronized (this) {
             leakedQueries = ImmutableSet.copyOf(leakedQueryReservations.keySet());
+            leakedBytes = leakedBytesThisTime;
         }
     }
 
-    private static boolean isLeaked(Map<QueryId, BasicQueryInfo> queryIdToInfo, QueryId queryId)
+    private static boolean isLeaked(Map<QueryId, Optional<BasicQueryInfo>> queryIdToInfo, QueryId queryId)
     {
-        BasicQueryInfo queryInfo = queryIdToInfo.get(queryId);
+        Optional<BasicQueryInfo> queryInfo = queryIdToInfo.get(queryId);
 
         if (queryInfo == null) {
             return true;
         }
 
-        DateTime queryEndTime = queryInfo.getQueryStats().getEndTime();
+        Optional<DateTime> queryEndTime = queryInfo.map(qi -> qi.getQueryStats().getEndTime());
 
-        if (queryInfo.getState() == RUNNING || queryEndTime == null) {
+        if (!queryEndTime.isPresent() || queryEndTime.get() == null || queryInfo.get().getState() == RUNNING) {
```

Comment:
```suggestion
        if (!queryEndTime.isPresent() || queryInfo.get().getState() == RUNNING) {
```

Optionals are guaranteed to never return null, so we can remove the null check.